batch/batch/server/server.py

```diff
@@ -458,7 +462,8 @@ def __init__(self, batch_id, job_id, attributes, callback, userdata, user, alway
         incomplete_parent_ids = await db.jobs.get_incomplete_parents(*self.id)
         if self._state == 'Pending' and not incomplete_parent_ids:
             parents = [Job.from_record(record) for record in await db.jobs.get_parents(*self.id)]
-            if self.always_run or all(p.is_successful() for p in parents):
+            if (self.always_run or
+                (all(p.is_successful() for p in parents) and not self._cancelled)):
```

Comment:
I would put self._cancelled first, short-circuit the linear time check on parents.